kinoml/modeling/alignment.py

```diff
@@ -0,0 +1,206 @@
+
+        # handle if target is KinaseDomainAminoAcidSequence vs. str
+        try:
+            sequence_id = getattr(target, "metadata")["uniprot_id"]
+            sequence_begin = getattr(target, "metadata")["begin"]
+            sequence_end = getattr(target, "metadata")["end"]
+        except:
+            sequence_id = "sequence_id"
+            sequence_begin = "1"
+            sequence_end = len(target)
+
+        protein_id = template.metadata["id"]
+        # get index for the first residue in the structure sequence
+        # ignoring negative and 0 resids
+        resids = template.universe.residues.resids
+        protein_start_index = next(
+            (i for i, x in enumerate(template.universe.residues.resids) if x > 0),
+            None,
+        )
+        protein_start = resids[protein_start_index]
+        protein_end = ""
+
+        if ligand:
+            # extract the AAs and No. dashes that need to be added in the alignment
+            tail_amino_acids, tail_dashes = self._format_alignment_with_ligands(
+                template, aligned_template_seq
+            )
+
+            # add on bufffer AAs / dashes so that Modeller finds the ligand correctly
+            seq1_dashed += tail_amino_acids
+            seq2_dashed += tail_dashes
+
+        # write alignment file in MODELLER format
+        with open(f"{self.alignment_file_path}", "w") as ali_file:
+            for i in range(len(seq1_dashed)):
+                if i == 0:
+                    ali_file.write(f">P1;{protein_id}\n")
+                    ali_file.write(
+                        f"structure:{protein_id}:{protein_start}:A:{protein_end}: :::     :     \n"
+                    )
+                ali_file.write(seq1_dashed[i])
+                if i == len(seq1_dashed) - 1:
+                    if ligand:
+                        ali_file.write("/.*")
+                    else:
+                        ali_file.write("*")
+                if (i + 1) % max_length == 0:
+                    ali_file.write("\n")
+
+            for i in range(len(seq2_dashed)):
+                # start new line below first sequence
+                if i == 0:
+                    ali_file.write(f"\n>P1;{sequence_id}\n")
+                    if ligand:
+                        ali_file.write(
+                            f"sequence:{sequence_id}:{sequence_begin}: :{sequence_end + len(tail_dashes)}: :::     :     \n"
+                        )
+                    else:
+                        ali_file.write(
+                            f"sequence:{sequence_id}:{sequence_begin}: :{sequence_end}: :::     :     \n"
+                        )
+                ali_file.write(seq2_dashed[i])
+                if i == len(seq2_dashed) - 1:
+                    if ligand:
+                        ali_file.write("/.*")
+                    else:
+                        ali_file.write("*")
+                if (i + 1) % max_length == 0:
+                    ali_file.write("\n")
```

Comment:
Have you checked this logic is not already covered in other libraries? I'd expect sequence alignment libraries like Biopython or Biotite to deal with this already, and they would have exposed their code to more use cases and robust testing.

Reply:
So Biotite is used to generate the alignments but I couldn't find a specific functionality in either Biotite or Biopython to write out `.ali` files in Modeller format.